mqtt/mqtt-broker/src/persist.rs

```diff
@@ -136,6 +145,211 @@ impl FileFormat for BincodeFormat {
+
+        let sessions = sessions
+            .into_iter()
+            .map(|s| this.consolidate_session(s))
+            .collect();
+
+        let payloads = this.payloads.into_iter().map(|(k, v)| (v, k));
+        let payloads = HashMap::from_iter(payloads);
+
+        ConsolidatedState {
+            payloads,
+            retained,
+            sessions,
+        }
+    }
+
+    fn consolidate_publication(&mut self, publication: Publication) -> SimplifiedPublication {
+        SimplifiedPublication {
+            topic_name: publication.topic_name,
+            retain: publication.retain,
+            qos: publication.qos,
+            payload: self.get_id(publication.payload),
+        }
+    }
+
+    fn consolidate_session(&mut self, session: SessionState) -> ConsolidatedSession {
+        let (client_id, subscriptions, waiting_to_be_sent) = session.into_parts();
+        let waiting_to_be_sent = waiting_to_be_sent
+            .into_iter()
+            .map(|p| self.consolidate_publication(p))
+            .collect();
+
+        ConsolidatedSession {
+            client_id,
+            subscriptions,
+            waiting_to_be_sent,
+        }
+    }
+
+    fn get_id(&mut self, bytes: Bytes) -> u64 {
+        if let Some(id) = self.payloads.get(&bytes) {
+            *id
+        } else {
+            self.curr_id += 1;
+            let id = self.curr_id;
+            self.payloads.insert(bytes, id);
+            id
+        }
+    }
+}
+
+struct Resolver {
+    payloads: HashMap<u64, Bytes>,
+}
+
+impl Resolver {
+    fn resolve_state(state: ConsolidatedState) -> BrokerState {
+        let mut this = Self {
+            payloads: state.payloads,
+        };
+
+        let retained = state
+            .retained
+            .into_iter()
+            .map(|(k, v)| (k, this.resolve_publication(v)));
+        let retained = HashMap::from_iter(retained);
+
+        let sessions = state
+            .sessions
+            .into_iter()
+            .map(|s| this.resolve_session(s))
+            .collect();
+
+        BrokerState::new(retained, sessions)
+    }
+
+    fn resolve_publication(&mut self, publication: SimplifiedPublication) -> Publication {
+        Publication {
+            topic_name: publication.topic_name,
+            retain: publication.retain,
+            qos: publication.qos,
+            payload: self.get_payload(publication.payload),
+        }
+    }
+
+    fn resolve_session(&mut self, session: ConsolidatedSession) -> SessionState {
+        let waiting_to_be_sent = session
+            .waiting_to_be_sent
+            .into_iter()
+            .map(|p| self.resolve_publication(p))
+            .collect();
+
+        SessionState::from_parts(session.client_id, session.subscriptions, waiting_to_be_sent)
+    }
+
+    fn get_payload(&mut self, id: u64) -> Bytes {
+        self.payloads
+            .get(&id)
+            .expect("All payloads should be stored.")
+            .clone()
```

Comment:
In this case clone is cheap but we should avoid cloning data. you can use remove method to move publication ownership from HashMap